File: apps/web/lib/crm.ts
// CRM client cho portal sale — types + gọi API agent-engine + helpers UI.
// Endpoint backend: app/api/crm.py (sale_router /sale/...).

import { AGENT_ENGINE_URL } from "./api";

// ---------------------------------------------------------------------------
// Types (đồng bộ với app/schemas/crm.py)
// ---------------------------------------------------------------------------

export type LeadSource =
  | "imported"
  | "registered"
  | "referral"
  | "fb_ads"
  | "zalo"
  | "email"
  | "manual"
  | "google_sheet"
  | "file_upload";

export type LeadStatus = "cold" | "warm" | "hot" | "customer" | "lost";

export type ContactChannel =
  | "call"
  | "sms"
  | "zalo"
  | "facebook"
  | "email"
  | "inperson";

export type ContactOutcome =
  | "no_answer"
  | "interested"
  | "not_interested"
  | "callback"
  | "booked";

export type CrmLead = {
  id: string;
  name: string;
  phone: string;
  email: string | null;
  source: LeadSource;
  status: LeadStatus;
  assigned_sale_id: string | null;
  imported_by_sale_id: string | null;
  ai_score: number;
  booking_count: number;
  contact_count: number;
  registered: boolean;
  last_contact_at: string | null;
  hot_marker_at: string | null;
  created_at: string;
  updated_at: string;
  note: string | null;
  days_since_contact: number | null;
};

export type ContactLog = {
  id: string;
  lead_id: string;
  sale_id: string;
  channel: ContactChannel;
  note: string;
  outcome: ContactOutcome;
  created_at: string;
};

export type CrmLeadDetail = CrmLead & {
  contact_logs: ContactLog[];
  assigned_sale_name: string | null;
};

export type LeadPage = {
  total: number;
  page: number;
  page_size: number;
  items: CrmLead[];
};

export type SaleTaskDaily = {
  sale_id: string;
  date: string;
  new_leads_added: number;
  contacts_made: number;
  meetings_attended: number;
  hot_leads_received: number;
  hot_leads_closed: number;
  score: number;
  target_new_leads: number;
  target_contacts: number;
  target_meetings: number;
  checked_in: boolean;
};

export type SalePerformance = {
  sale_id: string;
  sale_name: string;
  week_start: string;
  avg_daily_score: number;
  total_leads_added: number;
  total_hot_leads_received: number;
  total_deals_closed: number;
  eligibility_score: number;
  rank: number;
};

export type BulkImportResult = {
  imported: number;
  skipped: number;
  duplicates: { name: string; phone: string }[];
};

export type LeadInput = {
  name: string;
  phone: string;
  email?: string | null;
  note?: string | null;
  source?: LeadSource;
};

// ---- AI CRM (đồng bộ app/api/ai_crm.py) ----

export type AiTier = "cold" | "warm" | "hot";

export type AiNextAction = {
  summary?: string | null;
  suggested_action?: string | null;
};

export type LeadInsight = {
  lead_id: string;
  ai_score: number;
  ai_tier?: AiTier | string | null;
  ai_reason?: string | null;
  ai_best_time?: string | null;
  ai_next_action?: AiNextAction | null;
  ai_scored_at?: string | null;
  status?: string | null;
};

// ---------------------------------------------------------------------------
// Fetch wrapper (tự gắn Bearer, parse lỗi tiếng Việt)
// ---------------------------------------------------------------------------

async function req<T>(
  path: string,
  opts: { method?: string; token: string; body?: unknown } = { token: "" },
): Promise<T> {
  const res = await fetch(`${AGENT_ENGINE_URL}${path}`, {
    method: opts.method ?? "GET",
    headers: {
      ...(opts.token ? { Authorization: `Bearer ${opts.token}` } : {}),
      ...(opts.body !== undefined ? { "Content-Type": "application/json" } : {}),
    },
    body: opts.body !== undefined ? JSON.stringify(opts.body) : undefined,
    cache: "no-store",
  });
  const data = await res.json().catch(() => ({}) as Record<string, unknown>);
  if (!res.ok) {
    const detail = (data as { detail?: unknown }).detail;
    const message =
      typeof detail === "string"
        ? detail
        : Array.isArray(detail)
        ? detail
            .map((d: { msg?: string }) =>
              typeof d?.msg === "string" ? d.msg : JSON.stringify(d),
            )
            .join(", ")
        : `Lỗi ${res.status}`;
    throw new Error(message);
  }
  return data as T;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export function createLead(token: string, body: LeadInput): Promise<CrmLead> {
  return req<CrmLead>("/sale/leads", { method: "POST", token, body });
}

export function bulkImportLeads(
  token: string,
  leads: LeadInput[],
  skipDuplicates = true,
): Promise<BulkImportResult> {
  return req<BulkImportResult>("/sale/leads/bulk-import", {
    method: "POST",
    token,
    body: { leads, skip_duplicates: skipDuplicates },
  });
}

export function listMyLeads(
  token: string,
  params: { status?: string; search?: string; page?: number; page_size?: number } = {},
): Promise<LeadPage> {
  const qs = new URLSearchParams();
  if (params.status) qs.set("status", params.status);
  if (params.search) qs.set("search", params.search);
  if (params.page) qs.set("page", String(params.page));
  if (params.page_size) qs.set("page_size", String(params.page_size));
  const q = qs.toString();
  return req<LeadPage>(`/sale/leads${q ? `?${q}` : ""}`, { token });
}

export function getLeadDetail(token: string, id: string): Promise<CrmLeadDetail> {
  return req<CrmLeadDetail>(`/sale/leads/${id}`, { token });
}

export function updateLead(
  token: string,
  id: string,
  body: { status?: LeadStatus; note?: string; name?: string; phone?: string },
): Promise<CrmLead> {
  return req<CrmLead>(`/sale/leads/${id}`, { method: "PATCH", token, body });
}

export function addContactLog(
  token: string,
  id: string,
  body: { channel: ContactChannel; note: string; outcome: ContactOutcome },
): Promise<ContactLog> {
  return req<ContactLog>(`/sale/leads/${id}/contact-log`, {
    method: "POST",
    token,
    body,
  });
}

export function fetchTodayTask(token: string): Promise<SaleTaskDaily> {
  return req<SaleTaskDaily>("/sale/tasks/today", { token });
}

export function checkInToday(token: string): Promise<SaleTaskDaily> {
  return req<SaleTaskDaily>("/sale/tasks/check-in", { method: "POST", token });
}

export function fetchMyPerformance(token: string): Promise<SalePerformance> {
  return req<SalePerformance>("/sale/performance/me", { token });
}

export function fetchLeaderboard(token: string): Promise<SalePerformance[]> {
  return req<SalePerformance[]>("/sale/leaderboard", { token });
}

/** Insight AI 1 lead (tự chấm nếu chưa có / đã cũ). */
export function getLeadInsight(token: string, id: string): Promise<LeadInsight> {
  return req<LeadInsight>(`/ai-crm/leads/${id}/insight`, { token });
}

/** Chấm điểm lại 1 lead bằng AI (force). */
export function rescoreLead(token: string, id: string): Promise<LeadInsight> {
  return req<LeadInsight>(`/ai-crm/leads/${id}/rescore`, {
    method: "POST",
    token,
  });
}

// ---------------------------------------------------------------------------
// UI helpers — nhãn + màu badge tiếng Việt
// ---------------------------------------------------------------------------

export const STATUS_LABEL: Record<LeadStatus, string> = {
  cold: "Lạnh",
  warm: "Ấm",
  hot: "Nóng",
  customer: "Khách hàng",
  lost: "Đã mất",
};

export const STATUS_BADGE: Record<LeadStatus, string> = {
  cold: "bg-sky-50 text-sky-700 ring-sky-200",
  warm: "bg-amber-50 text-amber-700 ring-amber-200",
  hot: "bg-rose-50 text-rose-700 ring-rose-200",
  customer: "bg-emerald-50 text-emerald-700 ring-emerald-200",
  lost: "bg-brand-100 text-brand-600 ring-brand-200",
};

export const SOURCE_LABEL: Record<LeadSource, string> = {
  imported: "Danh bạ",
  registered: "Tự đăng ký",
  referral: "Giới thiệu",
  fb_ads: "FB Ads",
  zalo: "Zalo",
  email: "Email",
  manual: "Nhập tay",
  google_sheet: "Google Sheet",
  file_upload: "Tải file",
};

export const CHANNEL_LABEL: Record<ContactChannel, string> = {
  call: "Gọi điện",
  sms: "SMS",
  zalo: "Zalo",
  facebook: "Facebook",
  email: "Email",
  inperson: "Gặp trực tiếp",
};

export const OUTCOME_LABEL: Record<ContactOutcome, string> = {
  no_answer: "Không nghe máy",
  interested: "Quan tâm",
  not_interested: "Không quan tâm",
  callback: "Hẹn gọi lại",
  booked: "Đã đặt lịch",
};

export function scoreColor(score: number): string {
  if (score >= 70) return "text-rose-600";
  if (score >= 40) return "text-amber-600";
  return "text-sky-600";
}

export const TIER_LABEL: Record<string, string> = {
  cold: "Lạnh",
  warm: "Ấm",
  hot: "Nóng",
};

/** Màu badge cho tier AI (cold/warm/hot) — đồng bộ tông màu STATUS_BADGE. */
export const TIER_BADGE: Record<string, string> = {
  cold: "bg-sky-50 text-sky-700 ring-sky-200",
  warm: "bg-amber-50 text-amber-700 ring-amber-200",
  hot: "bg-rose-50 text-rose-700 ring-rose-200",
};

export function tierLabel(tier?: string | null): string {
  return TIER_LABEL[(tier ?? "").toLowerCase()] ?? "Chưa xếp";
}

export function tierBadge(tier?: string | null): string {
  return (
    TIER_BADGE[(tier ?? "").toLowerCase()] ??
    "bg-brand-100 text-brand-600 ring-brand-200"
  );
}

export function formatDate(iso: string | null): string {
  if (!iso) return "—";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return "—";
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${pad(d.getDate())}/${pad(d.getMonth() + 1)}/${d.getFullYear()}`;
}

export function formatDateTime(iso: string | null): string {
  if (!iso) return "—";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return "—";
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${pad(d.getHours())}:${pad(d.getMinutes())} ${pad(d.getDate())}/${pad(
    d.getMonth() + 1,
  )}`;
}

File: apps/agent-engine/app/schemas/crm.py
"""Schema CRM — quản lý khách hàng, contact log, daily task & hiệu suất sale.

Đây là CRM "đầy đủ" (sale import danh bạ + admin master view + hot lead
distribution), TÁCH BIỆT với `app/schemas/lead.py` (lead nguồn chat in-memory
dùng cho luồng booking). Khái niệm `Lead` ở đây là khách hàng trong CRM.

Lưu trữ: JSON store interim (app/core/lead_store.py + sale_task_store.py).
Sau Sprint kế tiếp migrate PostgreSQL — giữ interface store để swap dễ.
"""

from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Literal, Optional

from pydantic import BaseModel, EmailStr, Field


class LeadSource(str, Enum):
    """Nguồn của khách hàng trong CRM."""

    IMPORTED = "imported"  # sale import từ danh bạ
    REGISTERED = "registered"  # khách tự đăng ký landing
    REFERRAL = "referral"  # sale share link giới thiệu
    FB_ADS = "fb_ads"
    ZALO = "zalo"
    EMAIL = "email"
    MANUAL = "manual"  # admin nhập tay
    GOOGLE_SHEET = "google_sheet"  # admin import từ Google Trang tính
    FILE_UPLOAD = "file_upload"  # admin import từ file CSV/XLSX


class LeadStatus(str, Enum):
    """Vòng đời khách hàng (soft-delete = LOST)."""

    COLD = "cold"
    WARM = "warm"
    HOT = "hot"
    CUSTOMER = "customer"  # đã đặt cọc / chốt deal
    LOST = "lost"


class ContactChannel(str, Enum):
    """Kênh sale liên hệ khách."""

    CALL = "call"
    SMS = "sms"
    ZALO = "zalo"
    FACEBOOK = "facebook"
    EMAIL = "email"
    INPERSON = "inperson"
    NOTE = "note"  # ghi chú chăm sóc (không gắn kênh cụ thể)


ContactOutcome = Literal[
    "no_answer", "interested", "not_interested", "callback", "booked"
]


# ---------------------------------------------------------------------------
# Lead (khách hàng)
# ---------------------------------------------------------------------------

class LeadCreate(BaseModel):
    """Payload tạo 1 lead (sale nhập tay hoặc 1 dòng trong bulk import)."""

    name: str
    phone: str
    email: Optional[EmailStr] = None
    note: Optional[str] = None
    source: LeadSource = LeadSource.IMPORTED


class LeadBulkImport(BaseModel):
    """Payload import nhiều lead từ danh bạ (CSV / paste thô)."""

    leads: list[LeadCreate]
    skip_duplicates: bool = True


class LeadUpdate(BaseModel):
    """Payload cập nhật lead (sale: status/note)."""

    name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[EmailStr] = None
    status: Optional[LeadStatus] = None
    note: Optional[str] = None


class LeadAdminUpdate(BaseModel):
    """Payload SỬA thông tin khách (admin) — đầy đủ trường hồ sơ.

    Mọi field Optional: chỉ field gửi lên (exclude_unset) mới được cập nhật.
    `assigned_sale_id` cho phép admin đổi người phụ trách ngay tại form sửa.
    """

    name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[EmailStr] = None
    source: Optional[LeadSource] = None
    status: Optional[LeadStatus] = None
    note: Optional[str] = None
    assigned_sale_id: Optional[str] = None
    # Trường phân loại / hồ sơ mở rộng (Customer 360) — admin sửa được.
    region: Optional[str] = None
    customer_group: Optional[str] = None
    product_type: Optional[str] = None
    budget: Optional[str] = None
    purpose: Optional[str] = None
    project: Optional[str] = None


class Lead(BaseModel):
    """Khách hàng trong CRM. `days_since_contact` là computed cho UI."""

    id: str
    name: str
    phone: str
    email: Optional[str] = None
    source: LeadSource
    status: LeadStatus
    assigned_sale_id: Optional[str] = None  # ai đang phụ trách
    imported_by_sale_id: Optional[str] = None  # ai nhập đầu tiên
    ai_score: int = 0  # 0-100, AI tính dựa trên engagement
    booking_count: int = 0
    contact_count: int = 0
    registered: bool = False  # đã liên kết tài khoản web chưa
    last_contact_at: Optional[datetime] = None
    hot_marker_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    note: Optional[str] = None
    # Trường phân loại / hồ sơ mở rộng (Customer 360) — đều tuỳ chọn.
    region: Optional[str] = None          # Vùng miền / khu vực
    customer_group: Optional[str] = None  # Tệp khách / nhóm khách (nhãn tab import)
    product_type: Optional[str] = None    # Phân khúc / sản phẩm quan tâm
    budget: Optional[str] = None          # Ngân sách
    purpose: Optional[str] = None         # Mục đích (ở / đầu tư)
    project: Optional[str] = None         # Dự án quan tâm
    # Computed fields cho UI
    days_since_contact: Optional[int] = None


class LeadDetail(Lead):
    """Lead + lịch sử contact log (admin/sale detail view)."""

    contact_logs: list["ContactLog"] = Field(default_factory=list)
    assigned_sale_name: Optional[str] = None


# ---------------------------------------------------------------------------
# Contact log
# ---------------------------------------------------------------------------

class ContactLogCreate(BaseModel):
    channel: ContactChannel
    note: str
    outcome: ContactOutcome


class CareLogCreate(BaseModel):
    """Payload "đăng 1 hoạt động chăm sóc" lên dòng thời gian (care feed).

    Giống ContactLogCreate nhưng `outcome` TUỲ CHỌN (ghi chú thuần không cần kết
    quả) và `note` là nội dung bài đăng. created_by lấy từ user hiện tại ở endpoint.
    """

    channel: ContactChannel = ContactChannel.NOTE
    note: str
    outcome: Optional[ContactOutcome] = None


class ContactLog(BaseModel):
    id: str
    lead_id: str
    sale_id: str
    channel: ContactChannel
    note: str
    outcome: ContactOutcome
    created_by_name: Optional[str] = None
    created_at: datetime


# ---------------------------------------------------------------------------
# Sale daily task & performance
# ---------------------------------------------------------------------------

class SaleTaskDaily(BaseModel):
    sale_id: str
    date: date
    new_leads_added: int = 0  # số lead import hôm nay
    contacts_made: int = 0  # số contact log hôm nay
    meetings_attended: int = 0  # bookings completed hôm nay
    hot_leads_received: int = 0  # hot lead được phân bổ
    hot_leads_closed: int = 0  # chuyển sang customer
    score: int = 0  # 0-100, computed
    target_new_leads: int = 10
    target_contacts: int = 20
    target_meetings: int = 1
    checked_in: bool = False  # sale đã check-in chưa


class SalePerformance(BaseModel):
    sale_id: str
    sale_name: str
    week_start: date
    avg_daily_score: float
    total_leads_added: int
    total_hot_leads_received: int
    total_deals_closed: int
    eligibility_score: float  # dựa trên score → ưu tiên nhận hot lead
    rank: int


class SaleSuggestion(BaseModel):
    """Gợi ý sale khi PHÂN CÔNG chăm sóc 1 khách (trong hồ sơ 360°).

    Gộp hiệu suất (sale_task_store) + trạng thái online realtime (presence của
    Live Match) để admin chọn người MẠNH hoặc đang TRỰC. FE sắp xếp online + điểm
    cao lên trên (đã sort sẵn từ backend).
    """

    sale_id: str
    sale_name: str
    eligibility_score: float
    avg_daily_score: float
    total_deals_closed: int
    rank: int
    online: bool = False
    availability: Optional[str] = None  # online / busy / away / None
    active_calls: int = 0


# ---------------------------------------------------------------------------
# Admin dashboard stats
# ---------------------------------------------------------------------------

class CrmStats(BaseModel):
    total_leads: int
    hot_leads: int
    customers: int
    cold_leads: int
    warm_leads: int
    lost_leads: int
    conversion_rate: float  # customers / total (đã bỏ lost)
    top_sources: list[dict]  # [{source, count}]


LeadDetail.model_rebuild()
